### PR TITLE
[video_player] Validate size only when assets contain video tracks on iOS

### PR DESCRIPTION
**Pull request by AlexV525**
By adding a more specific type check, we can expand the support range of the video_player on iOS from videos to all supported media types. This is from the documentation of `AVSimplePlayer`: [AVSimplePlayer/AVSPDocument.m#139](https://developer.apple.com/library/archive/samplecode/AVSimplePlayerOSX/Listings/AVSimplePlayer_AVSPDocument_m.html#:~:text=tracksWithMediaType%3AAVMediaTypeVideo)

Resolves flutter/flutter#38480, flutter/flutter#90429, or more issues related to iOS audio playing.
Related to flutter/flutter#47105, flutter/flutter#75636.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by AlexV525** [comment]
Seems the duration for audios on iOS will always be `0:00:00.001000`. Drafting for further investigations.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m [hunk 1]
```diff
@@ -375,7 +377,7 @@ - (int64_t)position {
 }
 
 - (int64_t)duration {
-  return FLTCMTimeToMillis([[_player currentItem] duration]);
+  return FLTCMTimeToMillis([[[_player currentItem] asset] duration]);
```
Comment (AlexV525) [inline packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:380]: https://openradar.appspot.com/radar?id=4968600712511488

**Comment by AlexV525** [comment]
Hmm...1 millisecond of differences with the duration between platforms make tests fail.

File: packages/video_player/video_player/example/integration_test/video_player_test.dart [hunk 2]
```diff
@@ -240,4 +240,73 @@ void main() {
...
+    testWidgets('can be initialized', (WidgetTester tester) async {
+      await _controller.initialize();
+
+      expect(_controller.value.isInitialized, true);
+      expect(_controller.value.position, const Duration(seconds: 0));
+      expect(_controller.value.isPlaying, false);
+      // Due to the duration calculation accurancy between platforms,
+      // the milliseconds on Web will be a slightly different from natives.
+      // The audio was made with 44100 Hz, 192 Kbps CBR, and 32 bits.
+      if (kIsWeb) {
```
Comment (AlexV525) [inline packages/video_player/video_player/example/integration_test/video_player_test.dart:258]: It's 5.041625 sec. The web takes `5.042` and natives take `5.041`.

Comment (AlexV525) [reply packages/video_player/video_player/example/integration_test/video_player_test.dart:258]: I've done some debug steps, and the root cause of this difference is from here:
https://github.com/flutter/plugins/blob/a10ac2e0d8a8cd878fd0d0a54aa9a87c44c7d7c5/packages/video_player/video_player_web/lib/video_player_web.dart#L301
If we want to make them equal, we might use `toInt()` here.

**Comment by stuartmorgan** [comment]
I don't think this actually resolves flutter/flutter#47105 as claimed in the PR description; using `video_player` to play audio isn't actually a good idea, and will be even less so when if (most likely when) we migrate it to platform views.

I'm conflicted about this PR. The change itself is simple, but adding tests asserting that audio works sets a problematic precedent. What if a new platform we want to support has different APIs for supporting video and audio; are we now saying we require audio support in that implementation too? When there are inevitably new issues or feature requests that are specific to audio, will those all suddenly be valid?

I think where I come down on this is that because the mismatch with Android is currently a source of confusion, and the incremental cost is extremely low here, I'm inclined to take this PR, but:
- The official position should still be that we don't officially support audio
- We should still do https://github.com/flutter/flutter/issues/75636, and explicitly clarify in the README that future versions and/or platforms may not support audio (I can do this in a follow-up)
- The test group being added here should have a comment saying that audio playback is tested to prevent *accidental* regression, but could be removed in the future if there's a technical reason to do so.

@jmagman Does that sound good to you?

**Comment by jmagman** [comment]
> @jmagman Does that sound good to you?

SGTM

**Comment by AlexV525** [comment]
> using video_player to play audio isn't actually a good idea

That's more like it used to be supported implicitly, but someday it becomes invalid. I'll update the above suggestions.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m (hunk 1 above)
Comment (jmagman) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:380]: You may want to put this link in a comment over this line.

**Review by jmagman — APPROVED** [review]
This LGTM with comment nit.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m (hunk 1 above)
Comment (AlexV525) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:383]: Done.

**Pull request merged**